lib/liquid/tags/for.rb

```diff
@@ -170,6 +114,74 @@ def strict_parse(markup)
 
     private
 
+    def collection_segment(context)
+      offsets = context.registers[:for] ||= Hash.new(0)
+
+      from = if @from == :continue
+        offsets[@name].to_i
+      else
+        context.evaluate(@from).to_i
+      end
+
+      collection = context.evaluate(@collection_name)
+      collection = collection.to_a if collection.is_a?(Range)
+
+      limit = context.evaluate(@limit)
+      to = limit ? limit.to_i + from : nil
+
+      segment = Utils.slice_collection(collection, from, to)
+      segment.reverse! if @reversed
+
+      offsets[@name] = from + segment.length
+
+      segment
+    end
+
+    def render_segment(context, segment)
+      for_stack = context.registers[:for_stack] ||= []
+      length = segment.length
+
+      result = ''
+
+      context.stack do
+        loop_vars = {
+          'name'.freeze => @name,
+          'length'.freeze => length,
+          'parentloop'.freeze => for_stack[-1]
+        }
+
+        for_stack.push(loop_vars)
+
+        begin
+          context['forloop'.freeze] = loop_vars
+
+          segment.each_with_index do |item, index|
+            context[@variable_name] = item
+
+            loop_vars['index'.freeze] = index + 1
+            loop_vars['index0'.freeze] = index
+            loop_vars['rindex'.freeze] = length - index
+            loop_vars['rindex0'.freeze] = length - index - 1
+            loop_vars['first'.freeze] = (index == 0)
+            loop_vars['last'.freeze] = (index == length - 1)
```

Comment:
Should we instead use a liquid drop? Then we can just increment a single index between iterations of the loop, and calculate these values lazily on methods on the drop.